### DISCUSSION TITLE
Problem saving targets object to data/ in a user-defined R package

### DISCUSSION THREAD

**Discussion by luke-a-rogers** [ropensci/targets#588] [Help] [answered]
Hello! 

I absolutely love `targets`. I want to use it inside my own R packages where I like to run my analyses. I ran into a problem saving `targets` objects to `data/` via `usethis::use_data()`. Can you point me in the right direction to solve this problem?

Thanks so much!
Luke

## Motivation
I want to write my analysis as an R package, and save the output object to `data/` so that it is exported in a manner similar to other package data (mentioned [here](https://r-pkgs.org/data.html#data-data)).

## Problem
When I used the `usethis::use_data()` function inside my `write_to_data()` function, I got an error message:
```
> tar_make()
• start target my_data
• built target my_data
• start target write_my_data
...
x error target write_my_data
• end pipeline
Error : object ‘my_data’ not found
```

## Question
What is the preferred way to save a `targets` object to `data/` inside a user-defined R package?

**Comment by wlandau** [comment, marked as answer]
The difficulty comes from how `targets` handles in-memory data. To prevent accidental side-effects, in-memory targets are kept in a special environment that inherits from `tar_option_get("envir")`. To get around this, you can make sure each element of your data exists as a variable in the call frame of `write_to_data()`. 

```r
write_to_data <- function (...) {
  args <- list(...)
  envir <- environment()
  purrr::walk2(
    .x = names(args),
    .y = args,
    ~assign(x = .x, value = .y, envir = envir)
  )
  usethis::use_data(..., overwrite = TRUE)
  file.path("data", fs::path_ext_set(names(args)[1], ".rda"))
}
```

This assumes you name your arguments, e.g. `write_to_data(my_data = my_data)`.

I would also advise `tar_option_set(packages = "usedatargets", imports = "usedatargets")` instead of the older `tar_option_set(envir = getNamespace("usedatargets"))`: https://books.ropensci.org/targets/practices.html#packages-based-invalidation

**Reply by luke-a-rogers** [reply]
Wonderful! This works exactly as I hoped. Thank you so much!

**Reply by luke-a-rogers** [reply]
I got curious how to make sure my data existed as a variable in the call frame even when the argument was not named. I opted to have it throw an error if more than one object was passed to `write_to_data()`. Working through this helped me understand where `usethis::use_data()` looks to match the dots. This script worked for me, riffing off your ideas above. I'm curious if this may have any shortcomings that I haven't anticipated:

```r
write_to_data <- function (...) {
  if (...length() != 1) stop("write_to_data() takes exactly 1 argument")
  args <- list(...)
  if (is.null(names(args))) {
    name <- as.character(substitute(...))
  } else {
    name <- names(args)[1]
  }
  assign(x = name, value = ..1)
  usethis::use_data(..., overwrite = TRUE)
  file.path("data", fs::path_ext_set(name, ".rda"))
}
```

Thanks again!

**Reply by wlandau** [reply]
Nice, thanks for sharing.

If you only save one object, it may be simpler to do something like this. I have not tested it though.

```r
write_to_data <- function (value) {
  name <- deparse(substitute(value))
  assign(x = name, value = value)
  call <- as.call(
    list(
      quote(usethis::use_data),
      value = as.symbol(name),
      overwrite = TRUE
    )
  )
  eval(call)
  file.path("data", fs::path_ext_set(name, ".rda"))
}
```
